pulp_rpm/app/tasks/publishing.py

```diff
@@ -118,53 +119,57 @@ def publish_artifacts(self, content, prefix=""):
+        #
+        # Note that this only impacts user-created publications, which produce the "standard"
+        # RPM layout of repo/Packages/f/foo.rpm. A publication created by mirror-sync retains
+        # whatever layout their "upstream" repo-metadata dictates.
+        fields = ["pk", "relative_path", "content__rpm_package__time_build"]
+        for content_artifact in contentartifact_qs.values(*fields).iterator():
             relative_path = content_artifact["relative_path"]
+            time_build = content_artifact["content__rpm_package__time_build"]
+
             relative_path = os.path.join(
                 prefix, PACKAGES_DIRECTORY, relative_path.lower()[0], relative_path
             )
-            #
-            # Some Suboptimal Repos have the 'same' artifact living in multiple places.
-            # Specifically, the same NEVRA, in more than once place, **with different checksums**
-            # (since if all that was different was location_href there would be only one
-            # ContentArtifact in the first place).
-            #
-            # pulp_rpm wants to publish a 'canonical' repository-layout, under which an RPM
-            # "name-version-release-arch" appears at "Packages/n/name-version-release-arch.rpm".
-            # Because the assumption is that Packages don't "own" their path, only the filename
-            # is kept as relative_path.
-            #
-            # In this case, we have to pick one - which is essentially what the rest of the RPM
-            # Ecosystem does when faced with the impossible. This code takes the first-found. We
-            # could implement something more complicated, if there are better options
-            # (choose by last-created maybe?)
-            #
-            # Note that this only impacts user-created publications, which produce the "standard"
-            # RPM layout of repo/Packages/f/foo.rpm. A publication created by mirror-sync retains
-            # whatever layout their "upstream" repo-metadata dictates.
-            #
-            if relative_path in paths:
-                duplicated_paths.append(f'{relative_path}:{content_artifact["pk"]}')
-                continue
-            else:
-                paths.add(relative_path)
+            rel_path_mapping[relative_path].append((content_artifact["pk"], time_build))
+
+        for rel_path, content_artifacts in rel_path_mapping.items():
+            # sort the content artifacts by when the package was built
+            if len(content_artifacts) > 1:
+                content_artifacts.sort(key=lambda p: p[1], reverse=True)
+                log.warning(
+                    "Duplicate packages found competing for {path}, selected the one with "
+                    "the most recent build time, excluding {others} others.".format(
+                        path=rel_path, others=len(content_artifacts) - 1
+                    )
+                )
```

Comment:
Right now we have to have this logic in 2 places which sucks.  I definitely want to clean this up once it's not an emergency.